typing/typemod.ml

```diff
@@ -1449,262 +1540,299 @@ and type_module_aux ~alias sttn funct_body anchor env smod =
-(* TODO: check with Jacques why this is here
-      Tstr_class_type
-          (List.map (fun (_,_, i, d, _,_,_,_,_,_,c) -> (i, c)) classes) ::
-      Tstr_type
-          (List.map (fun (_,_,_,_, i, d, _,_,_,_,_) -> (i, d)) classes) ::
-      Tstr_type
-          (List.map (fun (_,_,_,_,_,_, i, d, _,_,_) -> (i, d)) classes) ::
-*)
-        List.flatten
-          (map_rec
-            (fun rs cls ->
-              let open Typeclass in
-              [Sig_class(cls.cls_id, cls.cls_decl, rs);
-               Sig_class_type(cls.cls_ty_id, cls.cls_ty_decl, rs);
-               Sig_type(cls.cls_obj_id, cls.cls_obj_abbr, rs);
-               Sig_type(cls.cls_typesharp_id, cls.cls_abbr, rs)])
-             classes []),
-        new_env
-    | Pstr_class_type cl ->
-        List.iter
-          (fun {pci_name} -> check_name check_type names pci_name)
-          cl;
-        let (classes, new_env) = Typeclass.class_type_declarations env cl in
-        Tstr_class_type
-          (List.map (fun cl ->
-               (cl.Typeclass.clsty_ty_id,
-                cl.Typeclass.clsty_id_loc,
-                cl.Typeclass.clsty_info)) classes),
-(*  TODO: check with Jacques why this is here
-           Tstr_type
-             (List.map (fun (_, _, i, d, _, _) -> (i, d)) classes) ::
-           Tstr_type
-             (List.map (fun (_, _, _, _, i, d) -> (i, d)) classes) :: *)
-        List.flatten
-          (map_rec
-             (fun rs decl ->
-                let open Typeclass in
-                [Sig_class_type(decl.clsty_ty_id, decl.clsty_ty_decl, rs);
-                 Sig_type(decl.clsty_obj_id, decl.clsty_obj_abbr, rs);
-                 Sig_type(decl.clsty_typesharp_id, decl.clsty_abbr, rs)])
-             classes []),
-        new_env
-    | Pstr_include sincl ->
-        let smodl = sincl.pincl_mod in
-        let modl =
-          Builtin_attributes.warning_scope sincl.pincl_attributes
-            (fun () -> type_module true funct_body None env smodl)
-        in
-        (* Rename all identifiers bound by this signature to avoid clashes *)
-        let sg = Subst.signature Subst.identity
-            (extract_sig_open env smodl.pmod_loc modl.mod_type) in
-        List.iter (check_sig_item names loc) sg;
-        let new_env = Env.add_signature sg env in
-        let incl =
-          { incl_mod = modl;
-            incl_type = sg;
-            incl_attributes = sincl.pincl_attributes;
-            incl_loc = sincl.pincl_loc;
-          }
-        in
-        Tstr_include incl, sg, new_env
-    | Pstr_extension (ext, _attrs) ->
-        raise (Error_forward (Builtin_attributes.error_of_extension ext))
-    | Pstr_attribute x ->
-        Builtin_attributes.warning_attribute x;
-        Tstr_attribute x, [], env
-  in
   let rec type_struct env sstr =
-    Ctype.init_def(Ident.current_time());
+    Ctype.init_def (Ident.current_time());
     match sstr with
     | [] -> ([], [], env)
-    | pstr :: srem ->
```

Comment:
It's a bit unfortunate to "break" the simple loop structure, because of the open case.   What about keeping the previous type_str_item for all other cases, and deal with Pstr_open as a special case in type_struct only?

Reply:
replied in https://github.com/ocaml/ocaml/pull/1506#discussion_r178452843